### PR TITLE
Add Ruby 2.7 support

### PR DESCRIPTION
**Pull request by wasifhossain**
#### Purpose
https://www.ruby-lang.org/en/news/2019/12/25/ruby-2-7-0-released/

Note: this PR is meant to experiment some cases that might arise while supporting Ruby 2.7, as it installs bundler 2.1.2 by default, which might cause dropping support for 
- Ruby 2.1/2.2
- Rails 4.1/4.2

### CODE REVIEW

**Comment by wasifhossain** [comment]
not sure why the last commit couldn't create a build on travis

**Comment by wasifhossain** [comment]
@bf4 for some reason, recent pull request events on travis are resulting in abuse. would you please look into the issue. thanks

image

if you are also on the same board as me, can we integrate circleci as a travis alternative? let me know if I can help.

**Comment by wasifhossain** [comment]
if the latest travis configuration runs fine, then it would confirm that we can keep supporting older ruby/rails versions

**Comment by bf4** [comment]
@wasifhossain you're welcome to set up a new CI if you'd like.  I find circle kind of annoying myself, though it's what I use at work now

**Comment by wasifhossain** [comment]
I would have preferred to keep using **travis**, but with the unknown reason for the above issue, we might have to try some other tools.

I have no preference for **circleci** too. would appreciate if you could set up some CI tool with AMS (or resolve the current issue), so that we can move forward without such issue

File: .travis.yml [hunk 1]
```diff
@@ -1,24 +1,25 @@
 language: ruby
-sudo: false
+os: linux
```
Comment (bf4) [inline .travis.yml]: why this change?

File: .travis.yml [hunk 2]
```diff
@@ -65,12 +70,13 @@ rvm:
...
   - *ruby_2_5
   - *ruby_2_6
+  - *ruby_2_7
   - *ruby_head
 
 branches:
   only: 0-10-stable
 
-matrix:
+jobs:
```
Comment (bf4) [inline .travis.yml]: this an api change in travis?

File: .travis.yml [hunk 3]
```diff
@@ -113,8 +123,13 @@ matrix:
     - { rvm: *ruby_2_4,  env: *rails_master }
 
   allow_failures:
+    # Rails 5.2.4.1 used safe-navigation operator (&.), which caused ruby 2.2 to fail
+    # https://buildkite.com/rails/rails/builds/65743#e3a74619-d8f6-493f-b590-1203028167d3
+    - { rvm: *ruby_2_2,  env: *rails_5_2 }
```
Comment (bf4) [inline .travis.yml]: instead of allowing failures we should exclude them

File: appveyor.yml [hunk 4]
```diff
@@ -14,9 +15,9 @@ cache:
 
 install:
   - SET PATH=C:\%ruby_version%\bin;%PATH%
-  - gem uninstall bundler -a -x -I
-  - gem update --system 2.7.9
-  - # gem install bundler -v '1.17.3'
+  - gem uninstall bundler -x
+  - # gem update --system 2.7.10
+  - gem install bundler -v 1.17.3
```
Comment (bf4) [inline appveyor.yml]: can you summarize what these changes address?

File: .travis.yml (hunk 1 above)
Comment (wasifhossain) [reply .travis.yml]: i was validating the contents of this file on https://config.travis-ci.com/explore, where I found suggestions to make these changes

File: lib/active_model/serializer.rb [hunk 5]
```diff
@@ -73,7 +73,7 @@ def self.serializer_lookup_chain_for(klass, namespace = nil)
     # Used to cache serializer name => serializer class
     # when looked up by Serializer.get_serializer_for.
     def self.serializers_cache
-      @serializers_cache ||= ThreadSafe::Cache.new
+      @serializers_cache ||= Concurrent::Map.new
```
Comment (bf4) [inline lib/active_model/serializer.rb]: seems out of scope

File: lib/active_model/serializer/reflection.rb [hunk 6]
```diff
@@ -87,8 +87,8 @@ def initialize(*)
       #       meta ids: ids
       #     end
       #   end
-      def link(name, value = nil)
-        options[:links][name] = block_given? ? Proc.new : value
+      def link(name, value = nil, &block)
+        options[:links][name] = block_given? ? block : value
```
Comment (bf4) [inline lib/active_model/serializer/reflection.rb]: seems out of scope

File: .travis.yml (hunk 2 above)
Comment (wasifhossain) [reply .travis.yml]: most probably. you may find it yourself by trying the link in my earlier comment

File: test/serializers/serializer_for_with_namespace_test.rb [hunk 7]
```diff
@@ -35,12 +35,11 @@ class PublisherSerializer < ActiveModel::Serializer
...
         attributes :title, :author_name
       end
+
       test 'resource without a namespace' do
         book = Book.new(title: 'A Post', author_name: 'hello')
 
-        # TODO: this should be able to pull up this serializer without explicitly specifying the serializer
-        #       currently, with no options, it still uses the Api::V3 serializer
-        result = ActiveModelSerializers::SerializableResource.new(book, serializer: BookSerializer).serializable_hash
+        result = ActiveModelSerializers::SerializableResource.new(book).serializable_hash
```
Comment (bf4) [inline test/serializers/serializer_for_with_namespace_test.rb]: seems this TODO was fixed at some point prior to this PR?

File: .travis.yml (hunk 3 above)
Comment (wasifhossain) [reply .travis.yml]: it seems like rails team will fix it in coming days (as rails 5-2-stable branch still supports ruby 2.2). so we may keep supporting it

File: .travis.yml (hunk 1 above)
Comment (bf4) [reply .travis.yml]: ok, I see 

```
[warn] on root: deprecated key: "ruby_supported_versions" (anchor on a non-private key)

[warn] on root: deprecated key: "jruby_supported_versions" (anchor on a non-private key)

[warn] on root: deprecated key: "jdk_supported_versions" (anchor on a non-private key)

[warn] on root: deprecated key: "rails_supported_versions" (anchor on a non-private key)

[warn] on root: deprecated key: "sudo" (The key `sudo` has no effect anymore.)

[info] on root: the key matrix is an alias for jobs, using jobs

[info] on env: the key matrix is an alias for jobs, using jobs

[info] on root: missing os, using the default "linux"

```

Is out of scope, I guess

File: appveyor.yml (hunk 4 above)
Comment (wasifhossain) [reply appveyor.yml]: in order to support ruby 2.1/2.2, we must ensure that bundler version stays below 2 and 1.17.3 is the latest version under 2.

in case bundler > 2 gets installed as part of rubygems update (which has been the case for rubygems 3+), we need to remove all installed bundler versions > 2 and then install the latest version < 2 explicitly

File: lib/active_model/serializer.rb (hunk 5 above)
Comment (wasifhossain) [reply lib/active_model/serializer.rb]: for reference: `thread_safe` gem is now deprecated and merged into `concurrent-ruby`

commit: https://github.com/ruby-concurrency/concurrent-ruby/commit/52e5f373a93abbe1f0339cf3087a269e4a76ff3e#diff-42d5a45da331eaa07d2b315bd3c9e738

Comment (wasifhossain) [reply lib/active_model/serializer.rb]: i will move this change to another PR if you like

File: lib/active_model/serializer/reflection.rb (hunk 6 above)
Comment (wasifhossain) [reply lib/active_model/serializer/reflection.rb]: ruby 2.7 makes separation of positional and keyword arguments, which throws deprecation warning. it will eventually be dropped in ruby 3, in which case current code is likely to break.

so i think we can keep this change

File: test/serializers/serializer_for_with_namespace_test.rb (hunk 7 above)
Comment (wasifhossain) [reply test/serializers/serializer_for_with_namespace_test.rb]: exactly

File: .travis.yml (hunk 1 above)
Comment (wasifhossain) [reply .travis.yml]: you mean that we move this change to another PR?

Comment (bf4) [reply .travis.yml]: yup.

File: lib/active_model/serializer/reflection.rb (hunk 6 above)
Comment (bf4) [reply lib/active_model/serializer/reflection.rb]: I'm at a loss to see the relationship between the argument changes in Ruby 2.7 and this diff. I believe that you saw a deprecation and that this made it go away.. but the reasoning sounds off

Let's get the 'out of scope' stuff extracted from this PR and merged before returning to this pr.

Comment (wasifhossain) [reply lib/active_model/serializer/reflection.rb]: ok. will do it soon

**Comment by bf4** [comment]
@wasifhossain I tried applying your changes after merging #2369

File: .travis.yml [hunk 8]
```diff
@@ -62,16 +65,17 @@ rvm:
...
 
 branches:
   only: 0-10-stable
 
-matrix:
+jobs:
   include:
+    - { rvm: *ruby_2_4,  env: *rails_4_2 }
+    - { rvm: *ruby_2_5,  env: *rails_4_2 }
+    - { rvm: *ruby_2_6,  env: *rails_4_2 }
```
Comment (wasifhossain) [inline .travis.yml]: afaik, these are already covered by the default configuration

File: .travis.yml [hunk 9]
```diff
@@ -93,9 +97,6 @@ matrix:
     - { rvm: *jruby_head, jdk: *jdk_8, env: *rails_master_jruby }
 
   exclude:
-    - { rvm: *ruby_2_4,  env: *rails_4_1 }
-    - { rvm: *ruby_2_5,  env: *rails_4_1 }
-    - { rvm: *ruby_2_6,  env: *rails_4_1 }
```
Comment (wasifhossain) [inline .travis.yml]: curious to know, you might have reasons to remove these?

**Comment by wasifhossain** [comment]
i think i have coded slightly different than yours

File: .travis.yml (hunk 9 above)
Comment (bf4) [reply .travis.yml]: I don't think they add much incremental value (I changed them from opt-out to opt-in)

Comment (bf4) [reply .travis.yml]: pushed again. diff was weird

**Comment by wasifhossain** [comment]
do you think rebasing this branch on 0-10-stable might help?

**Comment by wasifhossain** [comment]
whatever the reason, travis continues to deny starting a new build for the PR event :(

just found on [travis forum](https://travis-ci.community/t/abuse-detected-for-my-prs-and-some-commits/6124/13) that contacting about the denial issue to support@travis-ci.com might solve the case

can you please try this way. thanks

**Comment by wasifhossain** [comment]
could you please suggest next steps on how we could take it closer to merge. first we need to have travis allow us to run the matrix to make sure everything continues to work as usual

**Comment by bf4** [comment]
@wasifhossain it's green, it's it?

**Comment by wasifhossain** [comment]
checks are from **appveyor (ruby 2.3/rails 5.2) only**.

travis could **not** run due to the issue explained above

**Comment by wasifhossain** [comment]
image

File: .travis.yml (hunk 9 above)
Comment (wasifhossain) [reply .travis.yml]: looks like these lines are still missing from the new version. 

without these lines under the `exclude` section, default build matrix will try to run tests with these env and fail, eventually resulting in failure on travis

File: .travis.yml [hunk 10]
```diff
@@ -107,16 +106,12 @@ jobs:
     - { rvm: *ruby_2_1,  env: *rails_6_0 }
     - { rvm: *ruby_2_2,  env: *rails_6_0 }
     - { rvm: *ruby_2_3,  env: *rails_6_0 }
-    - { rvm: *ruby_2_4,  env: *rails_6_0 }
```
Comment (wasifhossain) [inline .travis.yml]: this line should be reverted back too

File: .travis.yml [hunk 11]
```diff
@@ -107,16 +106,12 @@ jobs:
...
 
     - { rvm: *ruby_2_1,  env: *rails_master }
     - { rvm: *ruby_2_2,  env: *rails_master }
     - { rvm: *ruby_2_3,  env: *rails_master }
-    - { rvm: *ruby_2_4,  env: *rails_master }
```
Comment (wasifhossain) [inline .travis.yml]: this line should be reverted back too

File: .travis.yml [hunk 12]
```diff
@@ -107,16 +106,12 @@ jobs:
...
 
   allow_failures:
-    - { rvm: *ruby_2_5,  env: *rails_master }
-    - { rvm: *ruby_2_6,  env: *rails_master }
```
Comment (wasifhossain) [inline .travis.yml]: though these lines are optional, reverting back would benefit more

**Comment by bf4** [comment]
@wasifhossain 

> checks are from appveyor (ruby 2.3/rails 5.2) only.

> travis could not run due to the issue explained above

oh, mea culpa. I didn't know that it wasn't even showing a build status. that's was unexpected to me

**Comment by bf4** [comment]
I just added travis as a required status for merging into 0-10-stable

btw, when I parse the travis.yml for the job config, I get

```ruby
 "env"=>
  {"global"=>["BUNDLER_VERSION=1.17.3"],
   "jobs"=>
    ["RAILS_VERSION=4.1",
     "RAILS_VERSION=4.2",
     "RAILS_VERSION=5.0",
     "RAILS_VERSION=5.1",
     "RAILS_VERSION=5.2",
     "RAILS_VERSION=6.0",
     "RAILS_VERSION=master"]},
 "branches"=>{"only"=>"0-10-stable"},
 "rvm"=>["2.1.10", "2.2.10", "2.3.8", "2.7.0", "ruby-head"],
 "jobs"=>
  {"include"=>
    [{"rvm"=>"2.4.9", "env"=>"RAILS_VERSION=4.2"},
     {"rvm"=>"2.5.7", "env"=>"RAILS_VERSION=4.2"},
     {"rvm"=>"2.6.5", "env"=>"RAILS_VERSION=4.2"},
     {"rvm"=>"jruby-9.1.17.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=4.1 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.1.17.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=4.2 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.1.17.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.0 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.1.17.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.1 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.1.17.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.2 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.2.8.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=4.2 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.2.8.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.0 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.2.8.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.1 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.2.8.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.2 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.2.8.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=6.0 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-9.2.8.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=master JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-head", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=4.2 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-head", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.0 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-head", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.1 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-head", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=5.2 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-head", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=6.0 JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-head", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=master JRUBY_OPTS='--dev -J-Xmx1024M --debug'"}],
   "exclude"=>
    [{"rvm"=>"ruby-head", "env"=>"RAILS_VERSION=4.1"},
     {"rvm"=>"2.1.10", "env"=>"RAILS_VERSION=5.0"},
     {"rvm"=>"2.1.10", "env"=>"RAILS_VERSION=5.1"},
     {"rvm"=>"2.1.10", "env"=>"RAILS_VERSION=5.2"},
     {"rvm"=>"2.1.10", "env"=>"RAILS_VERSION=6.0"},
     {"rvm"=>"2.2.10", "env"=>"RAILS_VERSION=6.0"},
     {"rvm"=>"2.3.8", "env"=>"RAILS_VERSION=6.0"},
     {"rvm"=>"2.1.10", "env"=>"RAILS_VERSION=master"},
     {"rvm"=>"2.2.10", "env"=>"RAILS_VERSION=master"},
     {"rvm"=>"2.3.8", "env"=>"RAILS_VERSION=master"}],
   "allow_failures"=>
    [{"rvm"=>"ruby-head"},
     {"rvm"=>"jruby-9.2.8.0", "jdk"=>"openjdk8", "env"=>"RAILS_VERSION=master JRUBY_OPTS='--dev -J-Xmx1024M --debug'"},
     {"rvm"=>"jruby-head"}],
   "fast_finish"=>true}}

```

**Comment by bf4** [comment]
sent travis an email

**Comment by wasifhossain** [comment]
> ok, so the forum shows I could link to https://travis-ci.org/rails-api/active_model_serializers/requests and it's there? how odd

you got it right :)

**Comment by bf4** [comment]
From travis

Jan 9, 02:50 EST

Hey Benjamin,

Thanks for your email and for your patience. Sorry for the confusion.

Our system automatically flagged your account for possible abuse, but after
further review, we were able to clear that up. You should be all set to
build as normal, though please feel free to contact us should you have any
questions or concerns arise.

Best,
Mustafa

**Comment by wasifhossain** [comment]
awesome! I will create a new PR with the proposed changes and request another review from you.

thanks!
